File: src/programmers/internal/json_schema_array.ts
import { OpenApi } from "@samchon/openapi";

import { MetadataArray } from "../../schemas/metadata/MetadataArray";

import { json_schema_plugin } from "./json_schema_plugin";
import { json_schema_station } from "./json_schema_station";

export const json_schema_array = (props: {
  components: OpenApi.IComponents;
  array: MetadataArray;
}): Array<OpenApi.IJsonSchema.IArray | OpenApi.IJsonSchema.IReference> => {
  const factory = (): OpenApi.IJsonSchema.IArray[] =>
    json_schema_plugin({
      schema: {
        type: "array",
        items: json_schema_station({
          blockNever: false,
          components: props.components,
          metadata: props.array.type.value,
          attribute: {},
        }),
      } satisfies OpenApi.IJsonSchema.IArray,
      tags: props.array.tags,
    });
  if (props.array.type.recursive === true) {
    const out = () => [
      {
        $ref: `#/components/schemas/${props.array.type.name}`,
      },
    ];
    if (props.components.schemas?.[props.array.type.name] !== undefined)
      return out();

    props.components.schemas ??= {};
    props.components.schemas[props.array.type.name] ??= {};

    const oneOf: OpenApi.IJsonSchema.IArray[] = factory();
    Object.assign(
      props.components.schemas[props.array.type.name]!,
      oneOf.length === 1 ? oneOf[0] : { oneOf },
    );
    return out();
  }
  return factory();
};

File: packages/typia/native/core/programmers/iterate/json_schema_array.go
package iterate

import nativemetadata "github.com/samchon/typia/packages/typia/native/core/schemas/metadata"

type Json_schema_array_export_props struct {
	Components *OpenApi_IComponents
	Array      *nativemetadata.MetadataArray
}

func json_schema_array(props struct {
	components *OpenApi_IComponents
	array      *nativemetadata.MetadataArray
}) []JsonSchema {
	factory := func() []JsonSchema {
		return json_schema_plugin(struct {
			schema JsonSchema
			tags   [][]nativemetadata.IMetadataTypeTag
		}{
			schema: JsonSchema{
				"type": "array",
				"items": json_schema_station(json_schema_station_props{
					blockNever: false,
					components: props.components,
					metadata:   props.array.Type.Value,
					attribute:  JsonSchema{},
				}),
			},
			tags: props.array.Tags,
		})
	}
	if props.array.Type.Recursive {
		out := func() []JsonSchema {
			return []JsonSchema{{"$ref": "#/components/schemas/" + props.array.Type.Name}}
		}
		if props.components.Schemas != nil {
			if _, ok := props.components.Schemas[props.array.Type.Name]; ok {
				return out()
			}
		}
		if props.components.Schemas == nil {
			props.components.Schemas = map[string]JsonSchema{}
		}
		if _, ok := props.components.Schemas[props.array.Type.Name]; ok == false {
			props.components.Schemas[props.array.Type.Name] = JsonSchema{}
		}
		oneOf := factory()
		if len(oneOf) == 1 {
			json_schema_assign(props.components.Schemas[props.array.Type.Name], oneOf[0])
		} else {
			props.components.Schemas[props.array.Type.Name]["oneOf"] = oneOf
		}
		return out()
	}
	return factory()
}

func Json_schema_array_export(props Json_schema_array_export_props) []JsonSchema {
	return json_schema_array(struct {
		components *OpenApi_IComponents
		array      *nativemetadata.MetadataArray
	}{
		components: props.Components,
		array:      props.Array,
	})
}
